linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -0,0 +1,3241 @@
+        fb += wa.Unreachable
+        NothingType
+      case _ =>
+        throw new IllegalArgumentException(
+          s"ArraySelect.array must be an array type, but has type ${t.array.tpe}"
+        )
+    }
+  }
+
+  private def genArrayValue(t: ArrayValue): Type = {
+    val arrayTypeRef = t.typeRef
+
+    markPosition(t)
+
+    genLoadVTableAndITableForArray(arrayTypeRef)
+
+    val expectedElemType = arrayTypeRef match {
+      case ArrayTypeRef(base: PrimRef, 1) => base.tpe
+      case _                              => AnyType
+    }
+
+    // Create the underlying array
+    t.elems.foreach(genTree(_, expectedElemType))
+    markPosition(t)
+    val underlyingArrayType = genTypeID.underlyingOf(arrayTypeRef)
+    fb += wa.ArrayNewFixed(underlyingArrayType, t.elems.size)
+
+    // Create the array object
+    fb += wa.StructNew(genTypeID.forArrayClass(arrayTypeRef))
+
+    t.tpe
+  }
+
+  private def genClosure(tree: Closure): Type = {
+    implicit val pos = tree.pos
+    implicit val ctx = this.ctx
+
+    val hasThis = !tree.arrow
+    val hasRestParam = tree.restParam.isDefined
+    val dataStructTypeID = ctx.getClosureDataStructType(tree.captureParams.map(_.ptpe))
+
+    // Define the function where captures are reified as a `__captureData` argument.
+    val closureFuncOrigName = genInnerFuncOriginalName()
+    val closureFuncID = new ClosureFunctionID(closureFuncOrigName)
+    emitFunction(
+      closureFuncID,
+      closureFuncOrigName,
+      enclosingClassName = None,
+      Some(tree.captureParams),
+      receiverType = if (!hasThis) None else Some(watpe.RefType.anyref),
+      tree.params,
+      tree.restParam,
+      tree.body,
+      resultType = AnyType
+    )
+
+    markPosition(tree)
+
+    // Put a reference to the function on the stack
+    fb += ctx.refFuncWithDeclaration(closureFuncID)
+
+    // Evaluate the capture values and instantiate the capture data struct
+    for ((param, value) <- tree.captureParams.zip(tree.captureValues))
+      genTree(value, param.ptpe)
+    markPosition(tree)
+    fb += wa.StructNew(dataStructTypeID)
+
+    /* If there is a ...rest param, the helper requires as third argument the
+     * number of regular arguments.
+     */
+    if (hasRestParam)
+      fb += wa.I32Const(tree.params.size)
+
+    // Call the appropriate helper
+    val helper = (hasThis, hasRestParam) match {
+      case (false, false) => genFunctionID.closure
+      case (true, false)  => genFunctionID.closureThis
+      case (false, true)  => genFunctionID.closureRest
+      case (true, true)   => genFunctionID.closureThisRest
+    }
+    fb += wa.Call(helper)
+
+    AnyType
+  }
+
+  private def genClone(t: Clone): Type = {
+    val expr = addSyntheticLocal(transformType(t.expr.tpe))
+
+    genTree(t.expr, ClassType(CloneableClass))
+
+    markPosition(t)
+
+    fb += wa.RefCast(watpe.RefType(genTypeID.ObjectStruct))
+    fb += wa.LocalTee(expr)
+    fb += wa.RefAsNonNull // cloneFunction argument is not nullable
```

Comment:
I do not understand why this instruction is necessary. Doesn't the `RefCast` already disallow `null`?